Path Author: Orson Teodoro <orsonteodoro@hotmail.com>
Date: Aug 20, 2020 (Unix timestamp 1597968203)
Subject:  Compile cargo packages libloading and glslopt with same ABI in
src_compile when set to CC=i686-pc-linux-gnu for target and
CC=x86_64-pc-linux-gnu for host.

Wrong detected compiler:
29:35.55      Running \
`/var/tmp/portage/www-client/firefox-79.0-r3/work/firefox-79.0-abi_x86_32.x86/\
ff/release/build/libloading-41e46a3f67cbb727/build-script-build`
29:35.62 [libloading 0.5.2] cargo:rustc-link-lib=dl
29:35.63 [libloading 0.5.2] TARGET = Some("x86_64-unknown-linux-gnu")
29:35.63 [libloading 0.5.2] OPT_LEVEL = Some("2")
29:35.63 [libloading 0.5.2] HOST = Some("x86_64-unknown-linux-gnu")
29:35.63 [libloading 0.5.2] CC_x86_64-unknown-linux-gnu = None
29:35.63 [libloading 0.5.2] CC_x86_64_unknown_linux_gnu = None
29:35.63 [libloading 0.5.2] HOST_CC = Some(\
"/usr/lib64/ccache/bin/x86_64-pc-linux-gnu-gcc -std=gnu99 -m32")
29:35.63 [libloading 0.5.2] CFLAGS_x86_64-unknown-linux-gnu = None
29:35.64 [libloading 0.5.2] CFLAGS_x86_64_unknown_linux_gnu = None
29:35.64 [libloading 0.5.2] HOST_CFLAGS = None
29:35.64 [libloading 0.5.2] CFLAGS = Some("-march=native -pipe")
29:35.64 [libloading 0.5.2] CRATE_CC_NO_DEFAULTS = None
29:35.64 [libloading 0.5.2] DEBUG = Some("false")
29:35.64 [libloading 0.5.2] lol target=x86_64-unknown-linux-gnu
29:35.64 [libloading 0.5.2] CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
29:35.64 [libloading 0.5.2] running: \
"/usr/lib64/ccache/bin/x86_64-pc-linux-gnu-gcc" "-std=gnu99" "-m32" "-O2" \
"-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-march=native" \
"-pipe" "-o" \
"/var/tmp/portage/www-client/firefox-79.0-r3/work/firefox-79.0-abi_x86_32.x86/\
ff/release/build/libloading-24c60df582c35fc3/out/src/os/unix/global_static.o

# Analysis
TARGET should be i686-unknown-linux-gnu
The line `running:` should contain -m32 ... -m32 NOT -m32 ... -m64
----
--- firefox-79.0/third_party/rust/cc/src/lib.rs.orig	2020-07-20 13:55:17.000000000 -0700
+++ firefox-79.0/third_party/rust/cc/src/lib.rs	2020-08-20 16:57:56.211827469 -0700
@@ -1435,7 +1435,14 @@ impl Build {
                 }
             }
             ToolFamily::Gnu => {
-                if target.contains("i686") || target.contains("i586") {
+                let host_cc = self.get_host_cc()?;
+                if host_cc.contains("-m32") {
+                    cmd.args.push("-m32".into());
+                } else if host_cc.contains("-mx32") {
+                    cmd.args.push("-mx32".into());
+                } else if host_cc.contains("-m64") {
+                    cmd.args.push("-m64".into());
+                } else if target.contains("i686") || target.contains("i586") {
                     cmd.args.push("-m32".into());
                 } else if target == "x86_64-unknown-linux-gnux32" {
                     cmd.args.push("-mx32".into());
@@ -2270,6 +2277,13 @@ impl Build {
         }
     }
 
+    fn get_host_cc(&self) -> Result<String, Error> {
+        match self.host.clone() {
+            Some(h) => Ok(h),
+            None => Ok(self.getenv_unwrap("HOST_CC")?),
+        }
+    }
+
     fn get_opt_level(&self) -> Result<String, Error> {
         match self.opt_level.as_ref().cloned() {
             Some(ol) => Ok(ol),
--- firefox-79.0/third_party/rust/cc/.cargo-checksum.json.orig	2020-08-20 13:47:50.260903607 -0700
+++ firefox-79.0/third_party/rust/cc/.cargo-checksum.json	2020-07-20 13:56:44.000000000 -0700
@@ -1 +1 @@
-{"files":{"Cargo.lock":"3aff5f8b0a7f4d72852b11b0526f0002e6bf55f19f1ebd6470d7f97fbd540e60","Cargo.toml":"6ab10d9b6a9c6f0909074e6698c90c6b6a7223661ec2e83174d2593117cbe7f2","LICENSE-APACHE":"a60eea817514531668d7e00765731449fe14d059d3249e0bc93b36de45f759f2","LICENSE-MIT":"378f5840b258e2779c39418f3f2d7b2ba96f1c7917dd6be0713f88305dbda397","README.md":"7184fbdf375a057e673257348f6d7584c0dd11b66318d98f3647f69eb610b097","src/bin/gcc-shim.rs":"b77907875029494b6288841c3aed2e4939ed40708c7f597fca5c9e2570490ca6","src/com.rs":"bcdaf1c28b71e6ef889c6b08d1ce9d7c0761344a677f523bc4c3cd297957f804","src/lib.rs":"4753929dbb7b676c19d7cfa06d0a47e37003554b80c536cbf2b892d591ef61c2","src/registry.rs":"3cc1b5a50879fa751572878ae1d0afbfc960c11665258492754b2c8bccb0ff5d","src/setup_config.rs":"7014103587d3382eac599cb76f016e2609b8140970861b2237982d1db24af265","src/winapi.rs":"ea8b7edbb9ff87957254f465c2334e714c5d6b3b19a8d757c48ea7ca0881c50c","src/windows_registry.rs":"388e79dcf3e84078ae0b086c6cdee9cf9eb7e3ffafdcbf3e2df26163661f5856","tests/cc_env.rs":"e02b3b0824ad039b47e4462c5ef6dbe6c824c28e7953af94a0f28f7b5158042e","tests/cflags.rs":"57f06eb5ce1557e5b4a032d0c4673e18fbe6f8d26c1deb153126e368b96b41b3","tests/cxxflags.rs":"c2c6c6d8a0d7146616fa1caed26876ee7bc9fcfffd525eb4743593cade5f3371","tests/support/mod.rs":"71620b178583b6e6e5e0d4cac14e2cef6afc62fb6841e0c72ed1784543abf8ac","tests/test.rs":"1605640c9b94a77f48fc92e1dc0485bdf1960da5626e2e00279e4703691656bc"},"package":"aa87058dce70a3ff5621797f1506cb837edd02ac4c0ae642b4542dce802908b8"}
\ No newline at end of file
+{"files":{"Cargo.lock":"3aff5f8b0a7f4d72852b11b0526f0002e6bf55f19f1ebd6470d7f97fbd540e60","Cargo.toml":"6ab10d9b6a9c6f0909074e6698c90c6b6a7223661ec2e83174d2593117cbe7f2","LICENSE-APACHE":"a60eea817514531668d7e00765731449fe14d059d3249e0bc93b36de45f759f2","LICENSE-MIT":"378f5840b258e2779c39418f3f2d7b2ba96f1c7917dd6be0713f88305dbda397","README.md":"7184fbdf375a057e673257348f6d7584c0dd11b66318d98f3647f69eb610b097","src/bin/gcc-shim.rs":"b77907875029494b6288841c3aed2e4939ed40708c7f597fca5c9e2570490ca6","src/com.rs":"bcdaf1c28b71e6ef889c6b08d1ce9d7c0761344a677f523bc4c3cd297957f804","src/lib.rs":"5c6125dd281da87956a7a636a18adb516eeed47cead7d8239c66a7a027d8ace7","src/registry.rs":"3cc1b5a50879fa751572878ae1d0afbfc960c11665258492754b2c8bccb0ff5d","src/setup_config.rs":"7014103587d3382eac599cb76f016e2609b8140970861b2237982d1db24af265","src/winapi.rs":"ea8b7edbb9ff87957254f465c2334e714c5d6b3b19a8d757c48ea7ca0881c50c","src/windows_registry.rs":"388e79dcf3e84078ae0b086c6cdee9cf9eb7e3ffafdcbf3e2df26163661f5856","tests/cc_env.rs":"e02b3b0824ad039b47e4462c5ef6dbe6c824c28e7953af94a0f28f7b5158042e","tests/cflags.rs":"57f06eb5ce1557e5b4a032d0c4673e18fbe6f8d26c1deb153126e368b96b41b3","tests/cxxflags.rs":"c2c6c6d8a0d7146616fa1caed26876ee7bc9fcfffd525eb4743593cade5f3371","tests/support/mod.rs":"71620b178583b6e6e5e0d4cac14e2cef6afc62fb6841e0c72ed1784543abf8ac","tests/test.rs":"1605640c9b94a77f48fc92e1dc0485bdf1960da5626e2e00279e4703691656bc"},"package":"aa87058dce70a3ff5621797f1506cb837edd02ac4c0ae642b4542dce802908b8"}
\ No newline at end of file
